File: src/sanitize.rs
// Content sanitization for rosie installs.
//
// Two defenses, exposed as a single options struct so callers can disable
// either independently:
//
//   - strip_invisible: remove zero-width, Unicode Tag block, and bidi-override
//     codepoints that the LLM reads but a human reviewer can't see in
//     rendered markdown.
//
//   - strip_comments: remove markdown comments outside fenced code blocks.
//     Two forms — HTML (`<!-- ... -->`, possibly multi-line) and link-form
//     (`[//]: # "..."` / `[//]: # (...)`).
//
// See docs/security and design/security.md for the threat model.

use crate::os;
use std::path::{Path, PathBuf};

#[derive(Debug, Clone, Copy)]
pub struct SanitizeOpts {
    pub strip_comments: bool,
    pub strip_invisible: bool,
}

impl SanitizeOpts {
    pub const ALL: SanitizeOpts = SanitizeOpts {
        strip_comments: true,
        strip_invisible: true,
    };

    pub const INVISIBLE_ONLY: SanitizeOpts = SanitizeOpts {
        strip_comments: false,
        strip_invisible: true,
    };

    pub const NONE: SanitizeOpts = SanitizeOpts {
        strip_comments: false,
        strip_invisible: false,
    };

    pub fn any(&self) -> bool {
        self.strip_comments || self.strip_invisible
    }
}

/// Sanitize reference content: strip comments and invisible chars per opts.
pub fn sanitize_reference(input: &str, opts: SanitizeOpts) -> String {
    let mut out = input.to_string();
    if opts.strip_comments {
        out = strip_comments(&out);
    }
    if opts.strip_invisible {
        out = strip_invisible(&out);
    }
    out
}

/// Sanitize skill content: strip invisible chars only (comments preserved —
/// skills are authored as agent input, their comments are intentional).
pub fn sanitize_skill(input: &str, opts: SanitizeOpts) -> String {
    if opts.strip_invisible {
        strip_invisible(input)
    } else {
        input.to_string()
    }
}

/// Walk `dir` recursively and rewrite every `.md` file with sanitize_skill.
/// Used after copy_dir_recursive to clean the canonical skill install.
pub fn sanitize_skill_dir(dir: &Path, opts: SanitizeOpts) -> os::Result<()> {
    if !opts.strip_invisible {
        return Ok(());
    }
    sanitize_skill_dir_inner(dir, opts)
}

fn sanitize_skill_dir_inner(dir: &Path, opts: SanitizeOpts) -> os::Result<()> {
    let entries = os::read_dir(dir)?;
    for name in entries {
        let path = PathBuf::from(dir).join(&name);
        if os::is_dir(&path) {
            sanitize_skill_dir_inner(&path, opts)?;
            continue;
        }
        if !is_markdown_file(&name) {
            continue;
        }
        let content = match os::read_to_string(&path) {
            Ok(c) => c,
            Err(_) => continue,
        };
        let cleaned = sanitize_skill(&content, opts);
        if cleaned != content {
            os::write(&path, cleaned.as_bytes())?;
        }
    }
    Ok(())
}

fn is_markdown_file(name: &str) -> bool {
    let lower = name.to_ascii_lowercase();
    lower.ends_with(".md") || lower.ends_with(".markdown")
}

// ---------------------------------------------------------------------------
// invisible-char stripping
// ---------------------------------------------------------------------------

fn strip_invisible(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    for (byte_idx, c) in s.char_indices() {
        if is_invisible(c, byte_idx) {
            continue;
        }
        out.push(c);
    }
    out
}

fn is_invisible(c: char, byte_idx: usize) -> bool {
    match c {
        // Zero-width
        '\u{200B}' | '\u{200C}' | '\u{200D}' => true,
        // BOM is fine at the very start of the doc, hostile anywhere else
        '\u{FEFF}' if byte_idx != 0 => true,
        // Bidi overrides + isolates (Trojan Source class)
        '\u{202A}'..='\u{202E}' | '\u{2066}'..='\u{2069}' => true,
        // Unicode Tag block — invisible codepoints that encode arbitrary ASCII
        c if (c as u32) >= 0xE0000 && (c as u32) <= 0xE007F => true,
        _ => false,
    }
}

// ---------------------------------------------------------------------------
// markdown-comment stripping (outside fenced code blocks)
// ---------------------------------------------------------------------------

fn strip_comments(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    let mut in_fence = false;
    let mut in_html_comment = false;

    for line in s.split_inclusive('\n') {
        let trimmed = line.trim_start();

        // Fence toggle has priority. An open HTML comment that crosses a fence
        // boundary is pathological; we end the comment implicitly at the fence.
        if trimmed.starts_with("```") || trimmed.starts_with("~~~") {
            in_fence = !in_fence;
            in_html_comment = false;
            out.push_str(line);
            continue;
        }
        if in_fence {
            out.push_str(line);
            continue;
        }

        let processed = process_line(line, &mut in_html_comment);
        out.push_str(&processed);
    }

    out
}

fn process_line(line: &str, in_html_comment: &mut bool) -> String {
    // Strip HTML comments first (handles single-line and multi-line state).
    let stripped_html = strip_html_comments_on_line(line, in_html_comment);

    // After HTML stripping, check whether the line is a link-form comment.
    if is_link_form_comment_line(&stripped_html) {
        return String::new();
    }
    stripped_html
}

fn strip_html_comments_on_line(line: &str, in_html_comment: &mut bool) -> String {
    let chars: Vec<char> = line.chars().collect();
    let n = chars.len();
    let mut out = String::with_capacity(line.len());
    let mut i = 0;

    while i < n {
        if *in_html_comment {
            if i + 2 < n && chars[i] == '-' && chars[i + 1] == '-' && chars[i + 2] == '>' {
                *in_html_comment = false;
                i += 3;
                continue;
            }
            i += 1;
            continue;
        }

        if i + 3 < n
            && chars[i] == '<'
            && chars[i + 1] == '!'
            && chars[i + 2] == '-'
            && chars[i + 3] == '-'
        {
            // Search for closing --> on or after this position.
            let mut j = i + 4;
            let mut closed_at: Option<usize> = None;
            while j + 2 < n {
                if chars[j] == '-' && chars[j + 1] == '-' && chars[j + 2] == '>' {
                    closed_at = Some(j);
                    break;
                }
                j += 1;
            }
            // Also check the last 3 chars (loop above stops before n-2).
            if closed_at.is_none() && n >= 3 && i + 4 <= n - 3 {
                let last = n - 3;
                if last >= i + 4
                    && chars[last] == '-'
                    && chars[last + 1] == '-'
                    && chars[last + 2] == '>'
                {
                    closed_at = Some(last);
                }
            }
            if let Some(end) = closed_at {
                i = end + 3;
                continue;
            }
            // Unterminated: rest of line is inside the comment, continue on next line.
            *in_html_comment = true;
            // Preserve the trailing newline if there is one.
            if line.ends_with('\n') {
                out.push('\n');
            }
            return out;
        }

        out.push(chars[i]);
        i += 1;
    }

    out
}

fn is_link_form_comment_line(line: &str) -> bool {
    let trimmed = line.trim();
    if !trimmed.starts_with("[//]:") {
        return false;
    }
    let rest = trimmed[5..].trim_start();
    if !rest.starts_with('#') {
        return false;
    }
    let rest = rest[1..].trim_start();
    if rest.is_empty() {
        return true;
    }
    let first = rest.chars().next().unwrap();
    matches!(first, '"' | '\'' | '(')
}

// ---------------------------------------------------------------------------
// tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn strips_zero_width() {
        let input = "hello\u{200B}world\u{200C}foo\u{200D}bar";
        assert_eq!(strip_invisible(input), "helloworldfoobar");
    }

    #[test]
    fn strips_bidi_overrides() {
        let input = "ok\u{202E}reversed\u{202C}";
        assert_eq!(strip_invisible(input), "okreversed");
    }

    #[test]
    fn strips_unicode_tag_block() {
        let input = "visible\u{E0070}\u{E0061}\u{E0079}content";
        assert_eq!(strip_invisible(input), "visiblecontent");
    }

    #[test]
    fn preserves_leading_bom() {
        let input = "\u{FEFF}rest";
        assert_eq!(strip_invisible(input), "\u{FEFF}rest");
    }

    #[test]
    fn strips_nonleading_bom() {
        let input = "rest\u{FEFF}of\u{FEFF}doc";
        assert_eq!(strip_invisible(input), "restofdoc");
    }

    #[test]
    fn strips_html_comment_single_line() {
        let input = "before<!-- hidden -->after\n";
        assert_eq!(strip_comments(input), "beforeafter\n");
    }

    #[test]
    fn strips_html_comment_multi_line() {
        let input = "a\n<!--\nignored\nstill ignored\n-->\nb\n";
        // The intermediate lines become empty (content was inside the comment),
        // but their newlines are preserved as blank lines.
        let out = strip_comments(input);
        assert!(out.contains("a\n"));
        assert!(out.contains("b\n"));
        assert!(!out.contains("ignored"));
    }

    #[test]
    fn preserves_html_comment_inside_fence() {
        let input = "```\n<!-- preserved -->\n```\n";
        assert_eq!(strip_comments(input), input);
    }

    #[test]
    fn preserves_tilde_fence() {
        let input = "~~~\n<!-- preserved -->\n~~~\n";
        assert_eq!(strip_comments(input), input);
    }

    #[test]
    fn strips_link_form_quoted() {
        let input = "before\n[//]: # \"hidden\"\nafter\n";
        assert_eq!(strip_comments(input), "before\nafter\n");
    }

    #[test]
    fn strips_link_form_parenthesized() {
        let input = "[//]: # (hidden)\n";
        assert_eq!(strip_comments(input), "");
    }

    #[test]
    fn preserves_link_form_inside_fence() {
        let input = "```\n[//]: # \"preserved\"\n```\n";
        assert_eq!(strip_comments(input), input);
    }

    #[test]
    fn sanitize_skill_keeps_comments() {
        let input = "<!-- skill author wrote this -->\ntext";
        let out = sanitize_skill(input, SanitizeOpts::ALL);
        assert_eq!(out, input);
    }

    #[test]
    fn sanitize_reference_strips_both() {
        let input = "<!-- hide -->visible\u{200B}\n";
        let out = sanitize_reference(input, SanitizeOpts::ALL);
        assert_eq!(out, "visible\n");
    }

    #[test]
    fn sanitize_reference_respects_opts() {
        let input = "<!-- keep -->\u{200B}gone";
        let only_invisible = sanitize_reference(input, SanitizeOpts::INVISIBLE_ONLY);
        assert_eq!(only_invisible, "<!-- keep -->gone");
    }
}

File: npm/rosie-skills/src/sanitize.ts
// Content sanitization for rosie installs. Ported from src/sanitize.rs.
//
//   - stripInvisible: remove zero-width, Unicode Tag block, and bidi-override
//     codepoints the LLM reads but a human reviewer can't see.
//   - stripComments: remove markdown comments outside fenced code blocks.
//     HTML form (<!-- ... -->, possibly multi-line) and link form
//     ([//]: # "..." / [//]: # (...)).

import * as path from "node:path";
import * as os from "./os.js";

export interface SanitizeOpts {
  stripComments: boolean;
  stripInvisible: boolean;
}

export const SANITIZE_ALL: SanitizeOpts = { stripComments: true, stripInvisible: true };
export const SANITIZE_INVISIBLE_ONLY: SanitizeOpts = { stripComments: false, stripInvisible: true };
export const SANITIZE_NONE: SanitizeOpts = { stripComments: false, stripInvisible: false };

export function sanitizeAny(opts: SanitizeOpts): boolean {
  return opts.stripComments || opts.stripInvisible;
}

// Sanitize reference content: strip comments and invisible chars per opts.
export function sanitizeReference(input: string, opts: SanitizeOpts): string {
  let out = input;
  if (opts.stripComments) out = stripComments(out);
  if (opts.stripInvisible) out = stripInvisible(out);
  return out;
}

// Sanitize skill content: strip invisible chars only (comments preserved).
export function sanitizeSkill(input: string, opts: SanitizeOpts): string {
  return opts.stripInvisible ? stripInvisible(input) : input;
}

// Walk `dir` recursively and rewrite every .md file with sanitizeSkill.
export function sanitizeSkillDir(dir: string, opts: SanitizeOpts): void {
  if (!opts.stripInvisible) return;
  sanitizeSkillDirInner(dir, opts);
}

function sanitizeSkillDirInner(dir: string, opts: SanitizeOpts): void {
  let entries: string[];
  try {
    entries = os.readDir(dir);
  } catch {
    return;
  }
  for (const name of entries) {
    const p = path.join(dir, name);
    if (os.isDir(p)) {
      sanitizeSkillDirInner(p, opts);
      continue;
    }
    if (!isMarkdownFile(name)) continue;
    let content: string;
    try {
      content = os.readToString(p);
    } catch {
      continue;
    }
    const cleaned = sanitizeSkill(content, opts);
    if (cleaned !== content) os.write(p, cleaned);
  }
}

function isMarkdownFile(name: string): boolean {
  const lower = name.toLowerCase();
  return lower.endsWith(".md") || lower.endsWith(".markdown");
}

// ---- invisible-char stripping ----------------------------------------------

function stripInvisible(s: string): string {
  let out = "";
  let first = true;
  for (const c of s) {
    const cp = c.codePointAt(0) as number;
    if (!isInvisible(cp, first)) out += c;
    first = false;
  }
  return out;
}

function isInvisible(cp: number, isLeading: boolean): boolean {
  // Zero-width
  if (cp === 0x200b || cp === 0x200c || cp === 0x200d) return true;
  // BOM is fine at the very start of the doc, hostile anywhere else
  if (cp === 0xfeff && !isLeading) return true;
  // Bidi overrides + isolates (Trojan Source class)
  if ((cp >= 0x202a && cp <= 0x202e) || (cp >= 0x2066 && cp <= 0x2069)) return true;
  // Unicode Tag block
  if (cp >= 0xe0000 && cp <= 0xe007f) return true;
  return false;
}

// ---- markdown-comment stripping (outside fenced code blocks) ----------------

// split a string into lines, each retaining its trailing '\n' (mirrors Rust
// split_inclusive('\n')).
function splitInclusive(s: string): string[] {
  const lines: string[] = [];
  let start = 0;
  for (let i = 0; i < s.length; i++) {
    if (s[i] === "\n") {
      lines.push(s.slice(start, i + 1));
      start = i + 1;
    }
  }
  if (start < s.length) lines.push(s.slice(start));
  return lines;
}

function stripComments(s: string): string {
  let out = "";
  let inFence = false;
  const state = { inHtmlComment: false };

  for (const line of splitInclusive(s)) {
    const trimmed = line.replace(/^\s+/, "");
    // Fence toggle has priority.
    if (trimmed.startsWith("```") || trimmed.startsWith("~~~")) {
      inFence = !inFence;
      state.inHtmlComment = false;
      out += line;
      continue;
    }
    if (inFence) {
      out += line;
      continue;
    }
    out += processLine(line, state);
  }
  return out;
}

function processLine(line: string, state: { inHtmlComment: boolean }): string {
  const strippedHtml = stripHtmlCommentsOnLine(line, state);
  if (isLinkFormCommentLine(strippedHtml)) return "";
  return strippedHtml;
}

function stripHtmlCommentsOnLine(line: string, state: { inHtmlComment: boolean }): string {
  const chars = Array.from(line);
  const n = chars.length;
  let out = "";
  let i = 0;

  while (i < n) {
    if (state.inHtmlComment) {
      if (i + 2 < n && chars[i] === "-" && chars[i + 1] === "-" && chars[i + 2] === ">") {
        state.inHtmlComment = false;
        i += 3;
        continue;
      }
      i += 1;
      continue;
    }

    if (i + 3 < n && chars[i] === "<" && chars[i + 1] === "!" && chars[i + 2] === "-" && chars[i + 3] === "-") {
      // Search for closing --> on or after this position.
      let j = i + 4;
      let closedAt = -1;
      while (j + 2 < n) {
        if (chars[j] === "-" && chars[j + 1] === "-" && chars[j + 2] === ">") {
          closedAt = j;
          break;
        }
        j += 1;
      }
      // Also check the last 3 chars (loop above stops before n-2).
      if (closedAt === -1 && n >= 3 && i + 4 <= n - 3) {
        const last = n - 3;
        if (last >= i + 4 && chars[last] === "-" && chars[last + 1] === "-" && chars[last + 2] === ">") {
          closedAt = last;
        }
      }
      if (closedAt !== -1) {
        i = closedAt + 3;
        continue;
      }
      // Unterminated: rest of line is inside the comment, continue next line.
      state.inHtmlComment = true;
      if (line.endsWith("\n")) out += "\n";
      return out;
    }

    out += chars[i];
    i += 1;
  }

  return out;
}

function isLinkFormCommentLine(line: string): boolean {
  const trimmed = line.trim();
  if (!trimmed.startsWith("[//]:")) return false;
  let rest = trimmed.slice(5).replace(/^\s+/, "");
  if (!rest.startsWith("#")) return false;
  rest = rest.slice(1).replace(/^\s+/, "");
  if (rest.length === 0) return true;
  const first = rest[0];
  return first === '"' || first === "'" || first === "(";
}
